### ISSUE TITLE
[Phase 2] Unsupervised Anomaly Detection

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/fraud-detection-adaptive#4]
## Summary

Implement the unsupervised anomaly detection layer using Isolation Forest and One-Class SVM ensemble. This phase creates the first scoring layer that identifies anomalies without requiring labeled data.

**Parent Issue**: #1

## Objectives

- Implement Isolation Forest wrapper with auto-contamination
- Implement One-Class SVM wrapper tuned for extreme class imbalance
- Create weighted score fusion for ensemble
- Normalize all scores to [0, 1] range

## Dependencies

- Phase 0 must be complete (synthetic data generator)
- Phase 1 must be complete (feature pipeline)

## Task Checklist

- [ ] Implement `src/models/__init__.py` - package init
- [ ] Implement `src/models/isolation_forest.py` - IF wrapper
- [ ] Implement `src/models/one_class_svm.py` - OCSVM wrapper
- [ ] Implement `src/models/unsupervised_ensemble.py` - ensemble scorer
- [ ] Create `tests/test_unsupervised.py` - unit tests
- [ ] Performance benchmark: <50ms per transaction

### Isolation Forest Wrapper

📄 **File:** `src/models/isolation_forest.py`

```python
"""Isolation Forest wrapper for fraud detection."""

from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import MinMaxScaler
import numpy as np
from typing import Optional


class IsolationForestDetector:
    """
    Isolation Forest anomaly detector with score normalization.

    Isolation Forest works by randomly selecting features and splits,
    then measuring how quickly a point becomes isolated. Anomalies
    are isolated faster (fewer splits needed).
    """

    def __init__(
        self,
        n_estimators: int = 100,
        contamination: str = 'auto',
        random_state: int = 42
    ):
        """
        Args:
            n_estimators: Number of trees in the forest
            contamination: Expected proportion of outliers ('auto' recommended)
            random_state: Random seed for reproducibility
        """
        self.model = IsolationForest(
            n_estimators=n_estimators,
            contamination=contamination,
            random_state=random_state,
            n_jobs=-1  # Use all CPU cores
        )
        self.scaler = MinMaxScaler()
        self._is_fitted = False

    def fit(self, X: np.ndarray) -> 'IsolationForestDetector':
        """Fit the isolation forest on training data."""
        self.model.fit(X)

        # Fit scaler on training scores for normalization
        raw_scores = -self.model.decision_function(X)  # Negate so higher = more anomalous
        self.scaler.fit(raw_scores.reshape(-1, 1))

        self._is_fitted = True
        return self

    def score(self, X: np.ndarray) -> np.ndarray:
        """
        Return anomaly scores in [0, 1] range.

        Higher scores indicate more anomalous samples.
        """
        if not self._is_fitted:
            raise ValueError("Model must be fitted before scoring")

        # Get raw scores (more negative = more anomalous in sklearn)
        raw_scores = -self.model.decision_function(X)

        # Normalize to [0, 1]
        normalized = self.scaler.transform(raw_scores.reshape(-1, 1)).flatten()

        # Clip to ensure [0, 1] range
        return np.clip(normalized, 0, 1)

    def predict(self, X: np.ndarray, threshold: float = 0.5) -> np.ndarray:
        """Return binary predictions (1 = anomaly, 0 = normal)."""
        scores = self.score(X)
        return (scores >= threshold).astype(int)
```

### One-Class SVM Wrapper

📄 **File:** `src/models/one_class_svm.py`

```python
"""One-Class SVM wrapper for fraud detection."""

from sklearn.svm import OneClassSVM
from sklearn.preprocessing import MinMaxScaler, StandardScaler
import numpy as np
from typing import Optional


class OneClassSVMDetector:
    """
    One-Class SVM anomaly detector with score normalization.

    One-Class SVM learns a boundary around normal data.
    Points outside this boundary are considered anomalies.

    WARNING: OCSVM scales poorly with dataset size.
    For >10k samples, consider using SGDOneClassSVM instead.
    """

    def __init__(
        self,
        nu: float = 0.0001,  # Expected proportion of outliers (1:10000)
        kernel: str = 'rbf',
        gamma: str = 'scale',
        max_samples: int = 10000  # Subsample if too large
    ):
        """
        Args:
            nu: Upper bound on fraction of outliers (should match fraud rate)
            kernel: Kernel type ('rbf' recommended for fraud detection)
            gamma: Kernel coefficient ('scale' uses 1/(n_features * X.var()))
            max_samples: Maximum samples to use for training (memory limit)
        """
        self.nu = nu
        self.kernel = kernel
        self.gamma = gamma
        self.max_samples = max_samples

        self.model = OneClassSVM(nu=nu, kernel=kernel, gamma=gamma)
        self.feature_scaler = StandardScaler()  # OCSVM needs scaled features
        self.score_scaler = MinMaxScaler()
        self._is_fitted = False

    def fit(self, X: np.ndarray) -> 'OneClassSVMDetector':
        """Fit the one-class SVM on training data."""
        # Subsample if too large
        if len(X) > self.max_samples:
            indices = np.random.choice(len(X), self.max_samples, replace=False)
            X_train = X[indices]
        else:
            X_train = X

        # Scale features (critical for SVM)
        X_scaled = self.feature_scaler.fit_transform(X_train)

        # Fit model
        self.model.fit(X_scaled)

        # Fit score scaler
        raw_scores = -self.model.decision_function(X_scaled)
        self.score_scaler.fit(raw_scores.reshape(-1, 1))

        self._is_fitted = True
        return self

    def score(self, X: np.ndarray) -> np.ndarray:
        """
        Return anomaly scores in [0, 1] range.

        Higher scores indicate more anomalous samples.
        """
        if not self._is_fitted:
            raise ValueError("Model must be fitted before scoring")

        X_scaled = self.feature_scaler.transform(X)
        raw_scores = -self.model.decision_function(X_scaled)
        normalized = self.score_scaler.transform(raw_scores.reshape(-1, 1)).flatten()

        return np.clip(normalized, 0, 1)

    def predict(self, X: np.ndarray, threshold: float = 0.5) -> np.ndarray:
        """Return binary predictions (1 = anomaly, 0 = normal)."""
        scores = self.score(X)
        return (scores >= threshold).astype(int)
```

### Unsupervised Ensemble

📄 **File:** `src/models/unsupervised_ensemble.py`

```python
"""Unsupervised anomaly detection ensemble."""

import numpy as np
from typing import Dict, Optional

from src.models.isolation_forest import IsolationForestDetector
from src.models.one_class_svm import OneClassSVMDetector


class UnsupervisedEnsemble:
    """
    Ensemble of unsupervised anomaly detectors.

    Combines Isolation Forest and One-Class SVM with weighted averaging.
    IF is typically more robust, so default weight is higher.
    """

    def __init__(
        self,
        if_weight: float = 0.6,
        svm_weight: float = 0.4,
        if_params: Optional[dict] = None,
        svm_params: Optional[dict] = None
    ):
        """
        Args:
            if_weight: Weight for Isolation Forest scores
            svm_weight: Weight for One-Class SVM scores
            if_params: Parameters for IsolationForestDetector
            svm_params: Parameters for OneClassSVMDetector
        """
        if abs(if_weight + svm_weight - 1.0) > 1e-6:
            raise ValueError("Weights must sum to 1.0")

        self.if_weight = if_weight
        self.svm_weight = svm_weight

        self.if_detector = IsolationForestDetector(**(if_params or {}))
        self.svm_detector = OneClassSVMDetector(**(svm_params or {}))

        self._is_fitted = False

    def fit(self, X: np.ndarray) -> 'UnsupervisedEnsemble':
        """Fit both detectors on training data."""
        self.if_detector.fit(X)
        self.svm_detector.fit(X)
        self._is_fitted = True
        return self

    def score(self, X: np.ndarray) -> np.ndarray:
        """
        Return weighted ensemble anomaly scores in [0, 1].

        Higher scores indicate more anomalous samples.
        """
        if not self._is_fitted:
            raise ValueError("Ensemble must be fitted before scoring")

        if_scores = self.if_detector.score(X)
        svm_scores = self.svm_detector.score(X)

        ensemble_scores = self.if_weight * if_scores + self.svm_weight * svm_scores

        return ensemble_scores

    def score_detailed(self, X: np.ndarray) -> Dict[str, np.ndarray]:
        """Return individual detector scores along with ensemble score."""
        return {
            'isolation_forest': self.if_detector.score(X),
            'one_class_svm': self.svm_detector.score(X),
            'ensemble': self.score(X)
        }

    def predict(self, X: np.ndarray, threshold: float = 0.5) -> np.ndarray:
        """Return binary predictions (1 = anomaly, 0 = normal)."""
        scores = self.score(X)
        return (scores >= threshold).astype(int)
```

## Verification Commands

```bash
# Run unsupervised model tests
pytest tests/test_unsupervised.py -v

# Integration test
python -c "
from src.data.synthetic import generate_transactions
from src.features.pipeline import FeaturePipeline
from src.models.unsupervised_ensemble import UnsupervisedEnsemble

# Generate data
df = generate_transactions(10000)
pipeline = FeaturePipeline()
X = pipeline.fit_transform(df).values

# Fit and score
ensemble = UnsupervisedEnsemble()
ensemble.fit(X)
scores = ensemble.score(X)

print(f'Score range: [{scores.min():.3f}, {scores.max():.3f}]')
print(f'Mean fraud score: {scores[df.is_fraud == 1].mean():.3f}')
print(f'Mean legit score: {scores[df.is_fraud == 0].mean():.3f}')
"
```

## Definition of Done

- [ ] Ensemble returns scores in [0, 1] range for all inputs
- [ ] Known anomalies (synthetic frauds) score higher than median
- [ ] Inference time <50ms per transaction
- [ ] Individual detectors can be used standalone
- [ ] Weights are configurable and validated
- [ ] All tests pass

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/__init__.py` | - | Create | Package init |
| `src/models/isolation_forest.py` | - | Create | IF wrapper |
| `src/models/one_class_svm.py` | - | Create | OCSVM wrapper |
| `src/models/unsupervised_ensemble.py` | - | Create | Ensemble |
| `tests/test_unsupervised.py` | - | Create | Unit tests |

## Concepts to Understand

Before implementing, review:
- **Isolation Forest algorithm**: Anomalies are isolated with fewer splits
- **One-Class SVM**: Learns boundary around normal data
- **Decision function**: In sklearn, more negative = more anomalous
- **Score normalization**: MinMaxScaler to [0, 1] range
